ICE: panicked at 'assertion failed: !substs.has_escaping_bound_vars()', compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:1475:9

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code

```Rust
impl F {
    fn call() {
        <Self as Fn(&TestResult)>::call
    }
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.53.0-nightly (48691ea6e 2021-03-29)
binary: rustc
commit-hash: 48691ea6e639640f110b43e33d4aba1f07e7415c
commit-date: 2021-03-29
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

### Error output

```
error[E0412]: cannot find type `F` in this scope
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:6
   |
1  |   impl F {
   |        ^ help: a trait with a similar name exists: `Fn`

error[E0412]: cannot find type `TestResult` in this scope
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:22
  |
3 |         <Self as Fn(&TestResult)>::call
  |                      ^^^^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `perses_node_priority_with_dfs_delta_reduced_mutant`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | / impl F {
2 | |     fn call() {
3 | |         <Self as Fn(&TestResult)>::call
4 | |     }
5 | | }
  | |_^ consider adding a `main` function to `perses_node_priority_with_dfs_delta_reduced_mutant.rs`

error[E0658]: use of unstable library feature 'fn_traits'
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:9
  |
3 |         <Self as Fn(&TestResult)>::call
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(fn_traits)]` to the crate attributes to enable

error[E0229]: associated type bindings are not allowed here
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:18
  |
3 |         <Self as Fn(&TestResult)>::call
  |                  ^^^^^^^^^^^^^^^ associated type not allowed here

thread 'rustc' panicked at 'assertion failed: !substs.has_escaping_bound_vars()', compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:1475:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (48691ea6e 2021-03-29) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `<impl at perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1: 5:2>::call`
#1 [typeck_item_bodies] type-checking all item bodies
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0229, E0412, E0601, E0658.
For more information about an error, try `rustc --explain E0229`.
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0412]: cannot find type `F` in this scope
  --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:6
   |
1  |   impl F {
   |        ^ help: a trait with a similar name exists: `Fn`

error[E0412]: cannot find type `TestResult` in this scope
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:22
  |
3 |         <Self as Fn(&TestResult)>::call
  |                      ^^^^^^^^^^ not found in this scope

error[E0601]: `main` function not found in crate `perses_node_priority_with_dfs_delta_reduced_mutant`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | / impl F {
2 | |     fn call() {
3 | |         <Self as Fn(&TestResult)>::call
4 | |     }
5 | | }
  | |_^ consider adding a `main` function to `perses_node_priority_with_dfs_delta_reduced_mutant.rs`

error[E0658]: use of unstable library feature 'fn_traits'
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:9
  |
3 |         <Self as Fn(&TestResult)>::call
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #29625 <https://github.com/rust-lang/rust/issues/29625> for more information
  = help: add `#![feature(fn_traits)]` to the crate attributes to enable

error[E0229]: associated type bindings are not allowed here
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:3:18
  |
3 |         <Self as Fn(&TestResult)>::call
  |                  ^^^^^^^^^^^^^^^ associated type not allowed here

thread 'rustc' panicked at 'assertion failed: !substs.has_escaping_bound_vars()', compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:1475:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/48691ea6e639640f110b43e33d4aba1f07e7415c/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/48691ea6e639640f110b43e33d4aba1f07e7415c/library/core/src/panicking.rs:92:14
   2: core::panicking::panic
             at /rustc/48691ea6e639640f110b43e33d4aba1f07e7415c/library/core/src/panicking.rs:50:5
   3: rustc_typeck::check::fn_ctxt::_impl::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::instantiate_value_path
   4: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_kind
   5: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
   6: rustc_typeck::check::fn_ctxt::checks::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_block_with_expected
   7: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
   8: rustc_typeck::check::expr::<impl rustc_typeck::check::fn_ctxt::FnCtxt>::check_return_expr
   9: rustc_typeck::check::check::check_fn
  10: rustc_infer::infer::InferCtxtBuilder::enter
  11: rustc_typeck::check::typeck
  12: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  13: rustc_data_structures::stack::ensure_sufficient_stack
  14: rustc_query_system::query::plumbing::force_query_with_job
  15: rustc_query_system::query::plumbing::get_query_impl
  16: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck
  17: rustc_middle::ty::<impl rustc_middle::ty::context::TyCtxt>::par_body_owners
  18: rustc_typeck::check::typeck_item_bodies
  19: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  20: rustc_data_structures::stack::ensure_sufficient_stack
  21: rustc_query_system::query::plumbing::force_query_with_job
  22: rustc_query_system::query::plumbing::get_query_impl
  23: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::typeck_item_bodies
  24: rustc_session::utils::<impl rustc_session::session::Session>::time
  25: rustc_typeck::check_crate
  26: rustc_interface::passes::analysis
  27: rustc_query_system::dep_graph::graph::DepGraph<K>::with_task_impl
  28: rustc_data_structures::stack::ensure_sufficient_stack
  29: rustc_query_system::query::plumbing::force_query_with_job
  30: rustc_query_system::query::plumbing::get_query_impl
  31: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  32: rustc_interface::passes::QueryContext::enter
  33: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  34: rustc_span::with_source_map
  35: rustc_interface::interface::create_compiler_and_run
  36: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (48691ea6e 2021-03-29) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `<impl at perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1: 5:2>::call`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 5 previous errors

Some errors have detailed explanations: E0229, E0412, E0601, E0658.
For more information about an error, try `rustc --explain E0229`.
```

</p>
</details>

Slightly smaller:
```rust
#![feature(fn_traits)]
fn call() {
    <x as Fn(&usize)>::call
}
```
```
error[E0412]: cannot find type `x` in this scope
 --> src/lib.rs:3:6
  |
3 |     <x as Fn(&usize)>::call
  |      ^ not found in this scope

error[E0229]: associated type bindings are not allowed here
 --> src/lib.rs:3:11
  |
3 |     <x as Fn(&usize)>::call
  |           ^^^^^^^^^^ associated type not allowed here

thread 'rustc' panicked at 'assertion failed: !substs.has_escaping_bound_vars()', compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs:1475:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

--- a/compiler/rustc_resolve/src/late/lifetimes.rs
+++ b/compiler/rustc_resolve/src/late/lifetimes.rs
@@ -2688,15 +2688,14 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                 Scope::Binder { hir_id, .. } => {
                     break *hir_id;
                 }
-                Scope::Body { id, .. } => break id.hir_id,
                 Scope::ObjectLifetimeDefault { ref s, .. }
                 | Scope::Elision { ref s, .. }
                 | Scope::Supertrait { ref s, .. }
                 | Scope::TraitRefBoundary { ref s, .. } => {
                     scope = *s;
                 }
-                Scope::Root => {
-                    // See issue #83907. Just bail out from looking inside.
+                Scope::Root | Scope::Body { .. } => {
+                    // See issues #83907 and #83693. Just bail out from looking inside.
                     self.tcx.sess.delay_span_bug(
                         rustc_span::DUMMY_SP,
                         "In fn_like_elision without appropriate scope above",

--- a/src/test/ui/typeck/issue-83693.rs
+++ b/src/test/ui/typeck/issue-83693.rs
@@ -1,0 +1,19 @@
+// Regression test for the ICE described in #83693.
+
+#![feature(fn_traits)]
+#![crate_type="lib"]
+
+impl F {
+//~^ ERROR: cannot find type `F` in this scope [E0412]
+    fn call() {
+        <Self as Fn(&TestResult)>::call
+        //~^ ERROR: cannot find type `TestResult` in this scope [E0412]
+        //~| associated type bindings are not allowed here [E0229]
+    }
+}
+
+fn call() {
+    <x as Fn(&usize)>::call
+    //~^ ERROR: cannot find type `x` in this scope [E0412]
+    //~| ERROR: associated type bindings are not allowed here [E0229]
+}

--- a/src/test/ui/typeck/issue-83693.stderr
+++ b/src/test/ui/typeck/issue-83693.stderr
@@ -1,0 +1,39 @@
+error[E0412]: cannot find type `F` in this scope
+  --> $DIR/issue-83693.rs:6:6
+   |
+LL | impl F {
+   |      ^ help: a trait with a similar name exists: `Fn`
+   | 
+  ::: $SRC_DIR/core/src/ops/function.rs:LL:COL
+   |
+LL | pub trait Fn<Args>: FnMut<Args> {
+   | ------------------------------- similarly named trait `Fn` defined here
+
+error[E0412]: cannot find type `TestResult` in this scope
+  --> $DIR/issue-83693.rs:9:22
+   |
+LL |         <Self as Fn(&TestResult)>::call
+   |                      ^^^^^^^^^^ not found in this scope
+
+error[E0412]: cannot find type `x` in this scope
+  --> $DIR/issue-83693.rs:16:6
+   |
+LL |     <x as Fn(&usize)>::call
+   |      ^ not found in this scope
+
+error[E0229]: associated type bindings are not allowed here
+  --> $DIR/issue-83693.rs:9:18
+   |
+LL |         <Self as Fn(&TestResult)>::call
+   |                  ^^^^^^^^^^^^^^^ associated type not allowed here
+
+error[E0229]: associated type bindings are not allowed here
+  --> $DIR/issue-83693.rs:16:11
+   |
+LL |     <x as Fn(&usize)>::call
+   |           ^^^^^^^^^^ associated type not allowed here
+
+error: aborting due to 5 previous errors
+
+Some errors have detailed explanations: E0229, E0412.
+For more information about an error, try `rustc --explain E0229`.

--- a/src/test/ui/typeck/issue-84768.rs
+++ b/src/test/ui/typeck/issue-84768.rs
@@ -1,0 +1,10 @@
+// Regression test for the ICE described in #84768.
+
+#![feature(fn_traits)]
+#![crate_type="lib"]
+
+fn transform_mut<F>(f: F) where F: for<'b> FnOnce(&'b mut u8) {
+    <F as FnOnce(&mut u8)>::call_once(f, 1)
+    //~^ ERROR: associated type bindings are not allowed here [E0229]
+    //~| ERROR: mismatched types [E0308]
+}

--- a/src/test/ui/typeck/issue-84768.stderr
+++ b/src/test/ui/typeck/issue-84768.stderr
@@ -1,0 +1,19 @@
+error[E0229]: associated type bindings are not allowed here
+  --> $DIR/issue-84768.rs:7:11
+   |
+LL |     <F as FnOnce(&mut u8)>::call_once(f, 1)
+   |           ^^^^^^^^^^^^^^^ associated type not allowed here
+
+error[E0308]: mismatched types
+  --> $DIR/issue-84768.rs:7:42
+   |
+LL |     <F as FnOnce(&mut u8)>::call_once(f, 1)
+   |                                          ^ expected tuple, found integer
+   |
+   = note: expected tuple `(&mut u8,)`
+               found type `{integer}`
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0229, E0308.
+For more information about an error, try `rustc --explain E0229`.

